### ISSUE TITLE
shotover is a high performance proxy

### ISSUE THREAD

**Issue by rukai** [shotover/shotover-proxy#1507]
https://github.com/shotover/shotover-proxy/pull/1476 kicked off a bunch of work to refactor shotover in order to achieve its full performance potential.
I had a plan for future work but it was mostly in my head, so I really need a high level issue to track the high level goal and keep track of remaining tasks.

The remaining tasks in order:
- [ ] remove old invariants
- [ ] unify outgoing connection logic
   + currently cassandra has a whole lot of custom connection logic.
- [ ] outgoing connection logic returns immediately when any messages are received
     + when an outgoing response receives a message it should trigger a tokio notify that starts the transform chain
- [ ] batch outgoing messages https://github.com/shotover/shotover-proxy/issues/1471
- [ ] add backpressure
    + https://github.com/shotover/shotover-proxy/issues/1479
    
Once this is complete, shotover should:
* have higher throughput
* have lower latency
* not be blocked by slow cassandra responses.
* be resilient by exerting backpressure when large messages or large overall throughput is received.
